ntoskrnl/se/token.c

```diff
@@ -2119,138 +2374,82 @@ NtOpenThreadTokenEx(
-                }
-            }
-
-            InitializeObjectAttributes(&ObjectAttributes, NULL, HandleAttributes,
-                                       NULL, Dacl ? &SecurityDescriptor : NULL);
-
-            Status = SepDuplicateToken(Token, &ObjectAttributes, EffectiveOnly,
-                                       TokenImpersonation, ImpersonationLevel,
-                                       KernelMode, &NewToken);
-            if (!NT_SUCCESS(Status))
-            {
-                DPRINT1("NtOpenThreadTokenEx(): Failed to duplicate the token (Status 0x%lx)\n", Status);
-            }
-
-            ObReferenceObject(NewToken);
-            Status = ObInsertObject(NewToken, NULL, DesiredAccess, 0, NULL,
-                                    &hToken);
-            if (!NT_SUCCESS(Status))
-            {
-                DPRINT1("NtOpenThreadTokenEx(): Failed to insert the token object (Status 0x%lx)\n", Status);
-            }
-        }
-        else
-        {
-            DPRINT1("NtOpenThreadTokenEx(): Failed to impersonate token from DACL (Status 0x%lx)\n", Status);
-        }
-    }
-    else
-    {
-        Status = ObOpenObjectByPointer(Token, HandleAttributes,
-                                       NULL, DesiredAccess, SeTokenObjectType,
-                                       PreviousMode, &hToken);
-        if (!NT_SUCCESS(Status))
-        {
-            DPRINT1("NtOpenThreadTokenEx(): Failed to open the object (Status 0x%lx)\n", Status);
-        }
-    }
-
-    if (Dacl) ExFreePoolWithTag(Dacl, TAG_ACL);
-
+    /* Call the private function to do the job */
+    Status = SepOpenThreadToken(Thread,
+                                ThreadHandle,
+                                Token,
+                                DesiredAccess,
+                                HandleAttributes,
+                                EffectiveOnly,
+                                CopyOnOpen,
+                                ImpersonationLevel,
+                                PreviousMode,
+                                &hToken);
+
+    /* Restore the impersonation back if needed */
     if (RestoreImpersonation)
     {
         PsRestoreImpersonation(PsGetCurrentThread(), &ImpersonationState);
     }
 
-    ObDereferenceObject(Token);
-
-    if (NT_SUCCESS(Status) && CopyOnOpen)
+    if (!NT_SUCCESS(Status))
     {
-        Status = PsImpersonateClient(Thread, NewToken, FALSE, EffectiveOnly, ImpersonationLevel);
-        if (!NT_SUCCESS(Status))
-        {
-            DPRINT1("NtOpenThreadTokenEx(): Failed to impersonate the client (Status 0x%lx)\n", Status);
-        }
+        DPRINT1("Failed to open the thread's token (Status 0x%lx)\n", Status);
+        ObDereferenceObject(Token);
+        ObDereferenceObject(Thread);
+        return Status;
     }
 
-    if (NewToken) ObDereferenceObject(NewToken);
-
+    /* Dereference all the stuff and give the token handle to caller */
```

Comment:
You could move both those ObDereferenceObject calls before the `if (!NT_SUCCESS(Status))` above: this would allow to not duplicate them.